### PR TITLE
Basic shift from client to local and hosted client. Small increments …

### PR DESCRIPTION
**Pull request by RiddhiBagadiaa**
…only

### CODE REVIEW

**Comment by RiddhiBagadiaa** [comment]
The purpose of this branch is to clean up serving.py. This was done by breaking up training_set() and features() into smaller functions and getting rid of duplicate code. Serving_test.py was also modified to reflect these changes

File: client/src/featureform/serving.py [hunk 1]
```diff
@@ -143,112 +84,211 @@ def _host_features(self, features, entities):
...
-            return entity_row.values[0]
+
+    def get_features_for_entity(self, entity_id, entity_value, all_feature_df):
+        entity = all_feature_df.loc[all_feature_df[entity_id] == entity_value].copy()
+        entity.drop(columns=entity_id, inplace=True)
+        if len(entity.values) > 0:
+            return entity.values[0]
         else:
-            raise Exception("No matching entities for {}".format(entity_tuple))
+            raise Exception(f"No matching entities for {entity_id}: {entity_id}")
```
Comment (ahmadnazeri) [inline client/src/featureform/serving.py:272]: Is the second value suppose to be something else?

File: client/src/featureform/serving.py [hunk 2]
```diff
@@ -143,112 +84,211 @@ def _host_features(self, features, entities):
...
-        if entity_col not in df.columns:
-            raise KeyError(f"Entity column does not exist: {entity_col}")
-        if value_col not in df.columns:
-            raise KeyError(f"Value column does not exist: {value_col}")
-        if timestamp_column != "" and timestamp_column not in df.columns:
-            raise KeyError(f"Timestamp column does not exist: {timestamp_column}")
-        if timestamp_column != "":
-            df = df[[entity_col, value_col, timestamp_column]]
-            df[timestamp_column] = pd.to_datetime(df[timestamp_column])
+        if feature['source_entity'] not in df.columns:
```
Comment (ahmadnazeri) [inline client/src/featureform/serving.py:277]: We are using this check often. Do you think it might be better in it's own function?

**Review by ahmadnazeri — APPROVED** [review]
LGTM; just two questions

File: client/src/featureform/serving.py (hunk 1 above)
Comment (RiddhiBagadiaa) [reply client/src/featureform/serving.py:266]: Yup, changed it

File: client/src/featureform/serving.py [hunk 3]
```diff
@@ -15,122 +15,57 @@
...
+        host = host or os.getenv('FEATUREFORM_HOST')
+        if host is None:
+            raise ValueError(
+                'If not in local mode then `host` must be passed or the environment'
+                ' variable FEATUREFORM_HOST must be set.'
+            )
+        channel = self._create_channel(host, insecure, cert_path)
+        self._stub = serving_pb2_grpc.FeatureStub(channel)
+
+    def _create_channel(self, host, insecure, cert_path):
```
Comment (sdreyer) [inline client/src/featureform/serving.py:46]: Check tls.py in the src/featureform directory. I added functions to take care of this for all the clients so its more consistent

Comment (sdreyer) [reply client/src/featureform/serving.py:46]: Its being used in the resource client

**Review by sdreyer — APPROVED** [review]

**Pull request merged**